src/chainstate/stacks/miner.rs

```diff
@@ -56,6 +56,149 @@ use util::vrf::*;
+        let header_reader = chainstate.reopen()?;
+        let mut clarity_tx = chainstate.block_begin(&anchor_block_bhh, &anchor_block,
+                                                    &MINER_BLOCK_BURN_HEADER_HASH, &MINER_BLOCK_HEADER_HASH);
+        let anchor_block_height = 
+            StacksChainState::get_anchored_block_header_info(&header_reader.headers_db, &anchor_block_bhh, &anchor_block)?
+            .ok_or(Error::NoSuchBlockError)?
+            .block_height;
+
+        clarity_tx.reset_cost(initial_cost);
+        Ok(StacksMicroblockBuilder {
+            anchor_block,
+            anchor_block_bhh,
+            anchor_block_height,
+            clarity_tx: Some(clarity_tx),
+            header_reader,
+            prev_microblock_header: None,
+            considered: Some(HashSet::new()),
+        })
+    }
+
+    pub fn mine_next_microblock(&mut self,
+                                mem_pool: &MemPoolDB,
+                                miner_key: &Secp256k1PrivateKey,
+                                miner_pubkey_hash: &Hash160) -> Result<StacksMicroblock, Error> {
+        let mut txs_to_broadcast = vec![];
+
+        let mut clarity_tx = self.clarity_tx.take()
+            .expect("Microblock already open and processing");
+
+        let mut considered = self.considered.take()
+            .expect("Microblock already open and processing");
+
+        let result = mem_pool.iterate_candidates(
+            &self.anchor_block_bhh, &self.anchor_block, self.anchor_block_height, &mut self.header_reader,
+            |micro_txs| {
+                for mempool_tx in micro_txs.into_iter() {
+                    if considered.contains(&mempool_tx.metadata.txid) {
+                        continue;
+                    } else {
+                        considered.insert(mempool_tx.metadata.txid.clone());
+                    }
+                    if mempool_tx.tx.anchor_mode != TransactionAnchorMode::OffChainOnly && mempool_tx.tx.anchor_mode != TransactionAnchorMode::Any {
+                        continue;
+                    }
+                    match StacksChainState::process_transaction(&mut clarity_tx, &mempool_tx.tx) {
+                        Ok(_) => {
+                            txs_to_broadcast.push(mempool_tx.tx);
+                        },
+                        Err(e) => {
+                            match e {
+                                Error::CostOverflowError(cost_before, cost_after, total_budget) => { 
+                                    warn!("Transaction {} reached block cost {}; budget was {}", mempool_tx.tx.txid(), &cost_after, &total_budget);
+                                    clarity_tx.reset_cost(cost_before.clone());
+                                },
+                                _ => {
+                                    warn!("Error processing TX {}: {}", mempool_tx.tx.txid(), e);
```

Comment:
Should this path include the erred transaction if it was a runtime error, i.e., so the miner gets the fee but doesn't produce an invalid block?

Reply:
Probably not -- I think "runtime error handling" needs to be done within `process_transaction` (which handles aborting side-effects for transactions that cannot be included in a block at all), and that any transactions that will deduct fees (and be included in the block) shouldn't return an Err from `process_transaction`. I'm not positive that `process_transaction` is currently handling runtime exceptions correctly, but I think that's unrelated to this PR.